### PR TITLE
Preliminary registered-buffer collective support via Inductor

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138029
* #138028

```
NOTE [lowering-time collective optimization]

In collective communication libraries such as NCCL, every rank maintains
communication buffers that are remotely accessible by some peers. Depending
on the underlying transport, remote accessibility may be established via
mechanisms such as ib_reg_mr, CUDA P2P, or CUDA multicast. Typically, these
buffers are private to the communication library by default, and
communication ops copy user data in and out of these buffers.

To prevent these copies, an optimization commonly known as "user buffer
registration" can be employed. This allows direct establishment of remote
accessibility on user buffers, eliminating the need for copying. However,
this optimization introduces stringent usage requirements, which are
typically hard to satisfy without being intrusive to the user code:

- Establishing remote accessibility is expensive and often done ahead of
time. In such implementations, all ranks must agree on the set of allocations
used for every collective op. Failing to meet this requirement can
lead to runtime errors or even silent correctness issues.
- Even if the collective communication library supports gracefully falling
back to "unregistered" implementations, the fallback mechanism would nullify
the optimization.
- Some communication mechanisms impose stricter requirements than others. For
example, CUDA's multicast + multi-mem instructions require all ranks to agree
not only on the allocations used for every collective but also on the offsets
within these allocations.

To support all different mechanisms with optimal results, we aim to satisfy
the strictest requirement for this family of optimizations - we ensures that
every collective op invocation is guaranteed to operate on the same
allocation, at the same offset, in every iteration.

For eligible collective ops, we identify communication buffers at lowering
time and optionally choose to lower the op to a different kernel
(ommunication libraries like NCCL handle both registered and non-registered
buffers transparently within the same op, though some may require different
ops for different cases). Later, the codegen will perform "persistent
allocation" to satisfy the aforementioned constraints, and optionally,
perform buffer planning to optimize overall memory usage.
```

### Changes
- Created `comm_lowering.py` for the lowerings of `_c10d_functional` ops. This is to prevent cluttering `lowering.py` as we add more lowering-time collective optimizations. This PR moved the lowerings for `all_reduce` and `all_reduce_` to the file.
- Added `comm_buffer_type: Dict[str, str]` to `GraphLowering` to track whether a buffer is a comm buffer and the type of the comm buffer.
- Added codegen allocation support for comm buffers of type "symm_mem".
- Added support for auto-lowering `_c10d_functional.all_reduce_` to `symm_mem.one_shot_all_reduce`.
- Added an Inductor config for collective optimizations in general (`config._collective`).

### Limitation
Currently, each persistently allocated comm buffer is dedicated to a single callsite. This is not viable in terms of memory usage. However, this is a neccesary intermediate state before we tackle memory planning for comm buffers.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @XilunWu

### CODE REVIEW

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -464,6 +464,8 @@ def __init__(
         # Below field is related to printing debug intermediate tensor values info for debugging
         self.all_codegen_kernel_names: OrderedSet[str] = OrderedSet()
 
+        self.comm_buffer_type: Dict[str, str] = {}
```
Comment (Chillee) [inline torch/_inductor/graph.py:467]: I do wonder whether this should be on Layout or the graph... I think I have a weak preference for this to be a Layout subclass.

cc: @eellison and @jansel for opinions

File: torch/_inductor/comm_lowering.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    assert isinstance(buffer, ir.Buffer), type(buffer)
+
+    # The buffer is already realized as a comm buffer
+    if buffer.get_name() in V.graph.comm_buffer_type:
+        return comm_buffer_type == V.graph.comm_buffer_type[buffer.get_name()]
+
+    # We can realized a buffer as comm buffer only if we know its size at
+    # codegen time.
+    try:
+        V.graph.sizevars.size_hint(buffer.get_numel())
```
Comment (Chillee) [inline torch/_inductor/comm_lowering.py:79]: This seems weird(?). Do you want to check whether it's all static shapes?

Comment (yifuwang) [reply torch/_inductor/comm_lowering.py:79]: Yup exactly. Please suggest a better way for doing it

File: torch/_inductor/graph.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/graph.py:467]: I agree a layout seems better.

File: torch/_inductor/comm_lowering.py (hunk 2 above)
Comment (Chillee) [reply torch/_inductor/comm_lowering.py:69]: We have a `is_symbolic` util: https://github.com/pytorch/pytorch/blob/e18439113e3e9260977e2bc52b83fa83a6fb30b0/torch/_inductor/fx_passes/pad_mm.py#L97

I'm actually not sure how your code even works. Shouldn't we always be able to get a size hint unless it's an unbacked symint?

File: test/distributed/test_symmetric_memory.py [hunk 3]
```diff
@@ -736,5 +737,76 @@ def _verify_all_reduce_result(self, inp, res):
...
+            return torch.ops._c10d_functional.wait_tensor(x)
+
+        compiled_0 = torch.compile(func_0, fullgraph=True)
+        code_0 = run_and_get_triton_code(compiled_0, arg)
+
+        self.assertIn("one_shot_all_reduce", code_0)
+        self.assertNotIn("return (buf0", code_0)
+
+        # All-reduce on a slice view
+        def func_1(x):
```
Comment (Chillee) [inline test/distributed/test_symmetric_memory.py:788]: Should add an example where the incoming is a matmul as well.

File: torch/_inductor/comm_lowering.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        )
+
+    buffer.layout = ir.CommBufferLayout(
+        layout=layout,
+        comm_buffer_type=comm_buffer_type,
+        group_name=group_name,
+    )
+
+
+_bufs_to_skip_wait: OrderedSet[Tuple[int, str]] = OrderedSet()
```
Comment (Chillee) [inline torch/_inductor/comm_lowering.py:116]: What is this for?

**Review by Chillee — CHANGES_REQUESTED** [review]
Mostly LGTM!

**Review by Chillee — APPROVED** [review]

**Comment by huydhn** [comment]
@yifuwang I think this change is failing on ROCm, could you help take a look at the failure?

distributed/test_symmetric_memory.py::LoweringTest::test_lowering_one_shot_all_reduce [GH job link](https://github.com/pytorch/pytorch/actions/runs/11601407475/job/32305535873) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/05cb98f91d49df9eadfcb3fc29bbd1b621d88860)

**Pull request closed without merging**